domain/application/types.go

```diff
@@ -178,8 +179,9 @@ type ContainerAddress struct {
 
 // AddUnitArg contains parameters for adding a unit to state.
 type AddUnitArg struct {
-	UnitName coreunit.Name
 	UnitStatusArg
+	UnitName    coreunit.Name
+	Constraints constraints.Constraints
```

Comment:
This was the correct way to do it. Instead of merging the constraints during AddUnit, they are passed by the callers.